### PR TITLE
V1/VR/V2 Disagree logic

### LINKED ISSUE #1774: [REQUEST]Add V1/VR/V2 Disagree Error Message to MCDU
**Is your feature request related to a problem? Please describe.**

**Describe the feature you'd like to see implemented**
<!-If the user tries to input any one of the takeoff speeds that does not meet the above condition, the MCDU should not accept the entry and should display the message, "V1/VR/V2 DISAGREE." -->

**References**

Screenshot (258)_LI

**Additional context**

Discord username (if different from GitHub):donbikes#4084

### PR DESCRIPTION
**Pull request by paul92ilm**
Fixes #1774 

## Summary of Changes

- [MCDU] Add V1/VR/V2 Disagree logic and message
- [ECAM] Add V1/VR/V2 Disagree test to T.O Config
- [ECAM] Add "T.O" ECAM Messages category 
- Fix minimum range for V1/VR/V2

## Screenshots (if necessary)

image
image

## References

[MCDU] V1/VR/V2 Logic
image

[MCDU] V1/VR/V2 Disagree message 
image

V1/VR/V2 range
image
image

[ECAM] T.O Config test
image

## Additional context

Discord username (if different from GitHub): Lussion

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Comment by lukecologne** [comment]
The ECAM warning seems to be <ins>T.O.</ins> V1/VR/V2 DISAGREE instead of <ins>CONFIG</ins> T.O. V1/VR/V2 DISAGREE. Source: 
grafik

**Comment by paul92ilm** [comment]
You're right, i'm fixing it

**Comment by paul92ilm** [comment]
Fixed
image

**Comment by lukecologne** [comment]
The warning should be only one line, not two

**Comment by paul92ilm** [comment]
image
image
Fixed, thanks for your feedback

File: .github/CHANGELOG.md [hunk 1]
```diff
@@ -35,13 +35,13 @@
...
 1. [MISC] Add localisation files for aircraft info when using other languages than english - @pessip (Pessi Päivärinne)
-1. [ATC] Fix CLR input for squawk code - @paul92ilm (Lussion)
+1. [ATC] Fixed CLR input for squawk code - @paul92ilm (Lussion)
 1. [TEXTURES] Fixed aft overhead circuit breaker decal from 'LIGHTNING' to 'LIGHTING' (@DarkOfNova#0001)
 1. [CAMERA] Modified a couple of the default interior camera angles. - @ZigTag (Falcon#5815), @linkeleven (linkeleven#2557)
 1. [CDU] Fix takeoff vspeed - @linkeleven (linkeleven#2557)
 1. [ECAM] Decimal font size change - @RichardPilbery (Richard Pilbery)
 1. [CDU] Add correct UI behavior and colors for airways page - @santii90 (Santiago Vazquez)
-1. [CDU] Add ability to clear MDA/DH and add input logic for both MDA/DH - @paul92ilm (Lussion)
+1. [CDU] Added ability to clear MDA/DH, added input logic for both MDA/DH - @paul92ilm (Lussion)
```
Comment (Benjozork) [inline .github/CHANGELOG.md:44]: Changelog lines start with action verbs, "Fix" and "Add" are fine.

**Review by Benjozork — CHANGES_REQUESTED** [review]

**Review by Benjozork — APPROVED** [review]

**Comment by Benjozork** [comment]
LGTM

**Comment by beheh** [comment]
Are you sure that the the entry is actually rejected into the MCDU? I see your code does not write the corresponding simvar if the condition does not match, showing just the error message and discarding the input. How would the T.O. check / ECAM message even be triggered if you cannot insert mismatching v speeds?

I would also imagine that a rejected entry leads to a cumbersome situation when updating V speeds - imagine pilots first doing the math for an intersection departure and entering something like 132 134 134, and then switching to full length departure redoing their calculations leading to 135 140 140. They would be unable to enter the 135 as the VR and V2 are smaller and would have to clear them first (is that even possible) or do them in reverse order.

You may want confirmation from a real pilot or at least someone with access to a professional simulator what exactly happens to the entry in the scratchpad.

**Comment by paul92ilm** [comment]
If TO data change, the "CHECK TAKE OFF DATA" message appears.
It allows to change V1/VR/V2 as follow without any problems (new issue on this point ?)
image

Here we deals with security so i'm not surprised about a double check.
ECAM check also allow to check UPLINK TO DATA I think
There is nothing in references saying values have to be kept or not

**Comment by beheh** [comment]
Sure, that screen is appropriate if the runway changes, but not if (for example) it starts to rain and the pilots decide to rerun the V speeds for the new runway conditions (which may also result in a new flex temperature).

This is not a big deal, but it seems ambiguous enough in the documentation that now we're adding it we might as well get a reference from a real aircraft/professional simulator rather than guessing one way or the other. I can easily see either way be correct, but we don't have to guess.

**Comment by paul92ilm** [comment]
Well, not so easy.. In phase 2 you have to push T.O Config but not in phase 3 (full page bellow)
image

ECAM logic message is not linked to the MCDU logic, so I agree with @beheh and wait for a pilot feedback

**Comment by HugoTro** [comment]
Category: QA Tester
Name: Hugo #3307
Date of testing: Current date 09/11/2020
Version of the sim: 1.10.8.0
PR Tested: #1786 
Object of test: V1,V2,V3 disagree logic on the MCDU

Testing technique: Cold & dark
Steps you've made to test the Pull Request:
1. Spawned in cold & dark at gate 37 at LFMN
2. Verified that we cannot enter V speeds if V1<VR<V2 


Issues : I can enter the same speeds for V1, Vr, and V2, which i tested with 90 knots
Overall Rating: Good
Conclusions of the testing: good PR !

**Comment by paul92ilm** [comment]
> Category: QA Tester
> Name: Hugo #3307
> Date of testing: Current date 09/11/2020
> Version of the sim: 1.10.8.0
> PR Tested: #1786
> Object of test: V1,V2,V3 disagree logic on the MCDU
> 
> Testing technique: Cold & dark
> Steps you've made to test the Pull Request:
> 
> 1. Spawned in cold & dark at gate 37 at LFMN
> 2. Verified that we cannot enter V speeds if V1<VR<V2
> 
> Issues : I can enter the same speeds for V1, Vr, and V2, which i tested with 90 knots
> Overall Rating: Good
> Conclusions of the testing: good PR !

image
Same V1, VR, V2 is allowed so everything normal

**Comment by HugoTro** [comment]
Ok cool i didn't see that 😃

**Comment by beheh** [comment]
Good news: norab (real pilot) from Discord found this after [a short discussion](https://discord.com/channels/738864299392630914/754851312306487326/775671983231991859) ([Source, page 3](https://safetyfirst.airbus.com/app/themes/mh_newsdesk/documents/archives/the-take-off-securing-function.pdf)):

image

While that isn't 100% the Honeywell documentation we usually use, I believe it's good enough of a reference to at least go by it for now until we find something better.

**Comment by beheh** [comment]
The [document from the previous post](https://safetyfirst.airbus.com/app/themes/mh_newsdesk/documents/archives/the-take-off-securing-function.pdf) is highly interesting - it describes additional consistency checks for V Speeds. These are not technically in scope, but could be picked up either here or in a future PR:

image

**Comment by paul92ilm** [comment]
Thanks @beheh, really interesting, I will change the code to keep values

Really ambiguous for the modified value but I think too that the message is displayed only when all 3 values have been filled

**Comment by ChruutvoLuzi** [comment]
Category: MCDU
Name: Discord User ChruutvoLuzi | SD-PB.ch#8902
Date of testing: Current date 12.11.2020
Version of the sim: 1.10.11.0
PR Tested: #1786 
Object of test: (VSpeeds Disagree Logic

Testing technique: Try out
Steps you've made to test the Pull Request:
1.  Spawned cold and dark
2.  Turned on the aircraft
3.  entered wrong VSpeeds
4. tested T.O Config Test Button in a few conditions (Taxi, Takeoff)

Flight notes (If required):

Issues (If Any): None
Overall Rating: Works as described
Conclusions of the testing: Noticed no issues. The warning is amber and triggers the master warning as described.

**Comment by FlightSimmer7700** [comment]
Category: QA Tester
Name: FlightSimmer7700#3468
Date of testing: Current date #12/11/20
Version of the sim: Latest
PR Tested: #1786 
Object of test: V Speeds

Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request:
1. Spawned at Jersey
2. Put in normal V speeds which worked as expected
3. Put in "incorrect" V speeds, Acted as described above
4.Incorrect v speeds comes up on the MCDU and upper ECAM when TO config button pressed

Flight notes (If required): 

Issues (If Any): none
Overall Rating: Good
Conclusions of the testing: Looks all good!

**Pull request merged**